x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/MlDailyMaintenanceServiceIT.java

```diff
@@ -0,0 +1,116 @@
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Consumer;
+
+import static java.util.stream.Collectors.toSet;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class MlDailyMaintenanceServiceIT extends MlNativeAutodetectIntegTestCase {
+
+    private JobConfigProvider jobConfigProvider;
+    private ThreadPool threadPool;
+
+    @Before
+    public void setUpMocks() {
+        jobConfigProvider = new JobConfigProvider(client(), xContentRegistry());
+        threadPool = mock(ThreadPool.class);
+        ExecutorService directExecutorService = EsExecutors.newDirectExecutorService();
+        when(threadPool.generic()).thenReturn(directExecutorService);
+    }
+
+    public void testTriggerDeleteJobsInStateDeletingWithoutDeletionTask() throws InterruptedException {
+        MlDailyMaintenanceService maintenanceService =
+            new MlDailyMaintenanceService(
+                settings(Version.CURRENT).build(),
+                ClusterName.DEFAULT,
+                threadPool,
+                client(),
+                mock(ClusterService.class),
+                mock(MlAssignmentNotifier.class));
+
+        putJob("maintenance-test-1");
+        putJob("maintenance-test-2");
+        putJob("maintenance-test-3");
+        assertThat(getJobIds(), containsInAnyOrder("maintenance-test-1", "maintenance-test-2", "maintenance-test-3"));
+
+        blockingCall(maintenanceService::triggerDeleteJobsInStateDeletingWithoutDeletionTask);
+        assertThat(getJobIds(), containsInAnyOrder("maintenance-test-1", "maintenance-test-2", "maintenance-test-3"));
+
+        this.<Boolean>blockingCall(listener -> jobConfigProvider.markJobAsDeleting("maintenance-test-2", listener));
+        this.<Boolean>blockingCall(listener -> jobConfigProvider.markJobAsDeleting("maintenance-test-3", listener));
+        assertThat(getJobIds(), containsInAnyOrder("maintenance-test-1", "maintenance-test-2", "maintenance-test-3"));
+        assertThat(getJob("maintenance-test-1").get(0).isDeleting(), is(false));
+        assertThat(getJob("maintenance-test-2").get(0).isDeleting(), is(true));
+        assertThat(getJob("maintenance-test-3").get(0).isDeleting(), is(true));
+
+        blockingCall(maintenanceService::triggerDeleteJobsInStateDeletingWithoutDeletionTask);
+        assertThat(getJobIds(), containsInAnyOrder("maintenance-test-1"));
+    }
+
+    private <T> void blockingCall(Consumer<ActionListener<T>> function) throws InterruptedException {
+        CountDownLatch latch = new CountDownLatch(1);
+        ActionListener<T> listener = ActionListener.wrap(
+            r -> {
+                latch.countDown();
+            },
+            e -> {
+                latch.countDown();
+                fail(e.getMessage());
```

Comment:
I'm not sure this will fail the test as the response does not come on the main test thread it will cause the assertion failure on a different thread. I could be wrong can you check please. 

Compare with https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/MlSingleNodeTestCase.java#L121

Reply:
I agree it's safer to assert on the main test thread so I changed it the way you suggested.

Now the code is very similar between my `blockingCall` and `MlSingleNodeTestCase::blockingCall`. Do you think it's worth extracting this method to some common place between single-node and multi-node tests?

Reply:
It is a simple method there is no harm in duplicating it.